src/components/profile-list.tsx

```diff
@@ -0,0 +1,328 @@
+import AnimeCover from "./anime-cover";
+import useSWR from "swr";
+import useSWRMutation from "swr/mutation";
+import React from "react";
+import Pagination from "./pagination";
+import RateDialog from "./rate-dialog";
+
+import { useRouter } from "next/router";
+import { z } from "zod";
+import { pageQuery, properCase } from "@/lib/utils";
+import { Button } from "./ui/button";
+import { ListType, PutListRequestSchema, listType } from "@/lib/types";
+import { List } from "@prisma/client";
+import { Loader2, MoreHorizontal } from "lucide-react";
+import { useMemo, useState } from "react";
+import Ratings, { Rating, ratingSchema } from "./ui/ratings";
+import { useSession } from "next-auth/react";
+import { DropdownMenu, DropdownMenuContent, DropdownMenuItem, DropdownMenuTrigger } from "./ui/dropdown-menu";
+import { DialogHeader, Dialog, DialogContent, DialogTitle } from "./ui/dialog";
+
+const fetcher = (url: string): Promise<{ list: Omit<List, "userId">[]; pages: number }> =>
+	fetch(url).then((res) => res.json());
+
+const fetcherMutate = (
+	url: string,
+
+	{ arg }: { arg: { listData: z.infer<typeof PutListRequestSchema>; method: "PUT" | "DELETE" } }
+) => {
+	return fetch(url, {
+		method: arg.method,
+		body: JSON.stringify({ listData: arg.listData }),
+		headers: {
+			"Content-Type": "application/json",
+		},
+	});
+};
+
+export default function ProfileList({ username }: { username: string }) {
+	const session = useSession();
+	const router = useRouter();
+	const [success, setSuccess] = useState("");
+	const [error, setError] = useState("");
+	const [rateDialog, setRateDialog] = useState<{
+		status: boolean;
+		id: string;
+		animeName: string;
+	}>({
+		id: "",
+		status: false,
+		animeName: "",
+	});
+
+	const key = useMemo(() => {
+		const listSchema = z.union([z.literal("watch"), z.literal("plan"), z.literal("rate")]).safeParse(router.query.list);
+		const pageSchema = z.number().safeParse(pageQuery());
+
+		if (listSchema.success && pageSchema.success) {
+			return `/api/list?username=${username}&list=${listSchema.data}&page=${pageSchema.data}`;
+		} else {
+			return null;
+		}
+	}, [router.query.list, router.query.page, username]);
+
+	const { data, isLoading, error: fetchError, mutate, isValidating } = useSWR(key, fetcher);
+	const { trigger } = useSWRMutation("/api/list", fetcherMutate);
+
+	const mutateList = async (
+		id: string,
+		animeName: string,
+		type: ListType | "update",
+		rate?: Rating,
+		ratedAt?: Date
+	) => {
+		// you should be able to delete request and put request
+		let res: Response | undefined;
+		const listRequestType = type === "watch" || type === "plan" ? type : (router.query.list as "watch" | "plan");
+		// means they want to delete the current list row
+		if (type === "delete") {
+			res = await trigger({ listData: { listRequestType, id }, method: "DELETE" });
+		}
```

Comment:
i understand it but this seems weird. You forcing typescript so this could be btter. Also, listRequestType isn't needed for the delete but you were probably doing to meet the type.

Reply:
cause this should realistically only have watch and plan as the type but we know do to the type of the parameter that isn't true.